File: src/nahyeon/BinarySearch/boj_10816_숫자카드_2.java
package nahyeon.BinarySearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class boj_10816_숫자카드_2 {

    static FastReader scan = new FastReader();
    static StringBuilder sb = new StringBuilder();

    static int N;
    static int[] NArr;

    static void input() {
        N = scan.nextInt();
        NArr = new int[N + 1];
        for (int i = 1; i <= N; i++) {
            NArr[i] = scan.nextInt();
        }
    }

    static int lower_bound(int[] A, int L, int R, int X) {
        int ans = R + 1;
        while (L <= R) {
            int mid = (L + R) / 2;
            if (A[mid] >= X) {
                ans = mid;
                R = mid - 1;
            } else {
                L = mid + 1;
            }
        }
        return ans;
    }

    static int upper_bound(int[] A, int L, int R, int X) {
        int ans = R + 1;
        while (L <= R) {
            int mid = (L + R) / 2;
            if (A[mid] > X) {
                ans = mid;
                R = mid - 1;
            } else {
                L = mid + 1;
            }
        }
        return ans;
    }

    static void pro() {
        // N 배열 정렬하기
        Arrays.sort(NArr, 1, N + 1);

        int M = scan.nextInt();
        for (int i = 1; i <= M; i++) {
            int M_elem = scan.nextInt();
            int upper = upper_bound(NArr, 1, N, M_elem);
            int lower = lower_bound(NArr, 1, N, M_elem);
            sb.append(upper - lower).append(' ');
        }
        System.out.println(sb);
    }

    public static void main(String[] args) {
        input();
        pro();
    }

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while (st == null || !st.hasMoreTokens()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }
    }
}

File: src/hyewon/BinarySearch/boj_10816_숫자 카드2.kt
package hyewon.BinarySearch

import java.io.*

fun lowerBound(arr: ArrayList<Int>, key:Int) : Int {
    var lo = 0
    var hi = arr.size

    while (lo < hi) {
        val mid = (lo + ((hi-lo) / 2))
        if (key <= arr[mid]) { //이상 조건
            hi = mid
        } else {
            lo = mid+1
        }
    }
    return lo
}
fun upperBound(arr : ArrayList<Int>, key:Int) : Int {
    var lo = 0
    var hi = arr.size

    while (lo < hi) {
        val mid = (lo + ((hi - lo)/2))
        if (key < arr[mid]) {
            hi = mid
        } else {
            lo = mid + 1
        }
    }
    return lo
}
fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()
    val arr = arrayListOf<Int>()
    val s = br.readLine().split(" ")
    for (i in 0 until n) {
        arr.add(s[i].toInt())
    }
    arr.sort()
    val n2 = br.readLine()!!.toInt()
    val s2 = br.readLine().split(" ")
    val sb =StringBuilder()
    for (j in 0 until n2) {
        val key = s2[j].toInt()
        sb.append(upperBound(arr,key) - lowerBound(arr,key)).append(" ")
    }
    println(sb)
}
